Pre-computed HLL support for hyperUnique aggregator

DaimonPl opened the pull request:
Final change for https://github.com/druid-io/druid/issues/3864

- addition of "isInputHyperUnique" parameter to hyperUnique aggregator
- addition of PreComputedHyperUniquesSerde which handles pre-computed ingestion
- hash function removed from serde and added as constant to HyperLogLogCollector (to ensure that both internal and external HLLs use same hash function)
- addition of ```#add(String rawValue)``` method to HyperLogLogCollector (so it's easier to add Strings)
- simple integration test added

gianm commented on hll/src/main/java/io/druid/hll/HyperLogLogCollector.java:
```diff
@@ -294,6 +298,11 @@ protected ByteBuffer getStorageBuffer()
     return storageBuffer;
   }
 
+  public void add(String rawValue)
```
The collector is supposed to be hash-function agnostic, so this is a bit out of place.

DaimonPl replied:
moved to HllHashing utility class

gianm commented on docs/content/querying/aggregations.md:
````diff
@@ -264,9 +264,16 @@ Determine the number of distinct starting characters of last names
+}
 ```
 
+isInputHyperUnique can be set to true to index pre-computed HLL (Base64 encoded output from druid-hll is expected).
````
Please include something mentioning that this is only used at ingestion time. Something like "The isInputHyperUnique field only affects ingestion-time behavior, and is ignored at query time."

DaimonPl replied:
added exactly as written :)

gianm commented on processing/src/main/java/io/druid/query/aggregation/hyperloglog/HyperUniquesSerde.java:
```diff
@@ -91,9 +80,7 @@ public HyperLogLogCollector extractValue(InputRow inputRow, String metricName)
-            collector.add(
-                hashFn.hashBytes(StringUtils.toUtf8(dimensionValue)).asBytes()
-            );
+            collector.add(dimensionValue);
```
What's the motivation behind removing the hashFn parameter of the serde? The collector is supposed to be hash-function agnostic, which is why the serde applied it before handing over the `hashedValue`.

DaimonPl replied:
moved to HllHashing utility class

gianm commented on processing/src/main/java/io/druid/query/aggregation/hyperloglog/PreComputedHyperUniquesSerde.java:
```diff
@@ -0,0 +1,74 @@
+  @Override
+  public String getTypeName()
+  {
+    return TYPE_NAME;
```
This is the type of the column that gets written out post-ingestion, so it should still be `hyperUnique`. (see also SketchMergeComplexMetricSerde, SketchBuildComplexMetricSerde).

DaimonPl replied:
done

DaimonPl commented:
Regarding motivation behind removing hash function from serde itself. I saw 2 problems:

1. (existed before my changes) from my understanding HLL cardinality estimation is based on hashes so if someone changes hashing algorithm then estimation between old-hashed and new-hashed values will be incorrect. In other words if druid changes hashing algorithm between versions older datasources won't be able to properly estimate values for old and new data together. Thus in my opinion hashing algorithm should be constant

2. (occurs after my changes) if we allow aggregation of pre-computed HLLs we should expect situation when someone changes aggregation of existing datasource from using isInputHyperUnique=false to isInputHyperUnique=true. In that case both old and new HLLs need to use exactly hashing algorithm and exactly same String to byte[] conversions (aka string encodings) because of reasons as in point "1"

I also thought about making hashing versionable but this would not solve problems described above as we would still end up with differently hashed values. That's why i moved hashing code to hll package and made it constant.

If my understanding is wrong please let me know and how to correct code :)

gianm commented:
@DaimonPl, you're right that the hashing algorithm can't just change at any time, because that would make merging older and newer HLLs impossible. But I think it still makes sense for it to be a parameter, for reasons of separation of concerns. There's no reason for the low level HLL code to have a particular hash function hard-coded. It makes sense for it to be something set up at the Druid server layer rather than the low-level `druid-hll` layer.

Also, not all HLL objects are long-lived. For example the cardinality aggregator creates them in the context of one query, so its hash function could change at any time and there would be no compatibility issues. Maybe that'd be worth doing if someone did some experimentation and found that another hash function gave better speed and/or accuracy.

If your main concern is making it easier for users of `druid-hll` to write HLLs compatible with Druid ingestion, how about keeping the hash function a parameter, but instead add a `HllHashing.defaultHashFunction()` method that returns the one that people should use for writing them to disk?

DaimonPl commented:
Updated code. Instead of static utility `HllHasing` I've added `HyperLogLogHash` class which takes `HashFunction` as constructor argument. So this way:
- `HyperUniquesSerde` can be parameterized with different `HashFunction` or whole `HyperLogLogHash`
- there's `HyperLogLogHash.createDefault()` so both internal and external hashing can be consistent
- `HyperLogLogHash` encapsulates `String` encoding so this is also consistent by default

let me know if now it's fine :)

gianm commented on processing/src/main/java/io/druid/query/aggregation/hyperloglog/PreComputedHyperUniquesSerde.java:
```diff
@@ -0,0 +1,72 @@
+        } else {
+          HyperLogLogCollector collector = HyperLogLogCollector.makeLatestCollector();
+
+          List<String> dimValues = inputRow.getDimension(metricName);
```
I'd suggest doing something similar to SketchMergeComplexMetricSerde here.

- use `inputRow.getRaw(metricName)` to get the actual value from the input row, not transformed into a `List<String>`
- handle some types it could be; see `SketchHolder.deserialize` for an example. For this one I'd suggest handling String, byte[], and HyperLogLogCollector.

This should be faster and should also work better with input rows that come from non-text sources, where the field might be a byte array or might already be a HLLC object. It'd also explicitly _not_ handle lists of HLLs which I think is not something that really makes sense, or that we'd want to support anyway.

DaimonPl replied:
Done, initial version was based on HyperUniquesSerde

gianm commented on hll/src/main/java/io/druid/hll/HyperLogLogHash.java:
```diff
@@ -0,0 +1,58 @@
+
+  public static HyperLogLogHash createDefault()
+  {
+    return new HyperLogLogHash(Hashing.murmur3_128());
```
This could be a singleton:

```java
private static final HyperLogLogHash DEFAULT = new HyperLogLogHash(Hashing.murmur3_128())
```

In which case I'd suggest renaming the method to `getDefault()` and just having it do `return DEFAULT;`.

DaimonPl replied:
Done :)

DaimonPl commented:
Will need to cleanup this git mess...

gianm commented:
@DaimonPl I'll take another look when it's cleaned up

DaimonPl commented:
cleaned to 1 single commit ;)

gianm commented:
@DaimonPl I raised #3932 to adjust formatting in PreComputedHyperUniquesSerde.java to match Druid standard style, in the future you can do this automatically by applying the IntelliJ or Eclipse code styles from here: https://github.com/druid-io/druid/blob/master/CONTRIBUTING.md
